File: src/com/intellij/idea/plugin/hybris/project/wizard/SelectHybrisModulesToImportStep.java
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.project.wizard;

import com.intellij.icons.AllIcons;
import com.intellij.ide.util.ElementsChooser;
import com.intellij.ide.util.projectWizard.WizardContext;
import com.intellij.idea.plugin.hybris.common.utils.HybrisIcons;
import com.intellij.idea.plugin.hybris.project.descriptors.*;
import com.intellij.idea.plugin.hybris.project.descriptors.impl.*;
import com.intellij.idea.plugin.hybris.settings.HybrisProjectSettings;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.ui.table.JBTable;
import org.apache.commons.lang3.BooleanUtils;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.intellij.idea.plugin.hybris.project.descriptors.ModuleDescriptorImportStatus.MANDATORY;
import static com.intellij.idea.plugin.hybris.project.descriptors.ModuleDescriptorImportStatus.UNUSED;

public class SelectHybrisModulesToImportStep extends AbstractSelectModulesToImportStep implements OpenSupport, RefreshSupport {

    private ModuleDescriptorImportStatus selectionMode = MANDATORY;

    public SelectHybrisModulesToImportStep(final WizardContext context) {
        super(context);
    }

    @Override
    protected void init() {
        this.fileChooser.addElementsMarkListener((ElementsChooser.ElementsMarkListener<ModuleDescriptor>) (element, isMarked) -> {
            if (element instanceof final YModuleDescriptor yModuleDescriptor) {
                if (isMarked) {
                    for (final ModuleDescriptor moduleDescriptor : yModuleDescriptor.getAllDependencies()) {
                        if (BooleanUtils.isNotFalse(fileChooser.getElementMarkStates().get(moduleDescriptor))) {
                            continue;
                        }

                        fileChooser.setElementMarked(moduleDescriptor, true);
                        if (selectionMode == MANDATORY) {
                            moduleDescriptor.setImportStatus(MANDATORY);
                        }
                    }
                }
                // Re-mark sub-modules accordingly
                markSubmodules(yModuleDescriptor, isMarked);
            }
            fileChooser.repaint();
        });
    }

    @Override
    public void updateStep() {
        getContext().setCoreStepModuleList();
        super.updateStep();
        selectionMode = MANDATORY;
        for (int index = 0; index < fileChooser.getElementCount(); index++) {
            final ModuleDescriptor yModuleDescriptor = fileChooser.getElementAt(index);
            if (yModuleDescriptor.isPreselected()) {
                fileChooser.setElementMarked(yModuleDescriptor, true);
                yModuleDescriptor.setImportStatus(MANDATORY);
            }
        }
        selectionMode = UNUSED;
        final Set<String> duplicateModules = new HashSet<>();
        final Set<String> uniqueModules = new HashSet<>();
        getContext().getList().forEach(e -> {
            if (uniqueModules.contains(e.getName())) {
                duplicateModules.add(e.getName());
            } else {
                uniqueModules.add(e.getName());
            }
        });

        // TODO: improve sorting
        fileChooser.sort((o1, o2) -> {
            final boolean o1dup = duplicateModules.contains(o1.getName());
            final boolean o2dup = duplicateModules.contains(o2.getName());
            if (o1dup ^ o2dup) {
                return o1dup ? -1 : 1;
            }

            final boolean o1custom = isCustomDescriptor(o1);
            final boolean o2custom = isCustomDescriptor(o2);
            if (o1custom ^ o2custom) {
                return o1custom ? -1 : 1;
            }

            // de-boost mandatory Platform extensions
            final boolean o1ext = isPlatformExtDescriptor(o1);
            final boolean o2ext = isPlatformExtDescriptor(o2);
            if (o1ext ^ o2ext) {
                return o2ext ? -1 : 1;
            }

            final boolean o1selected = isMandatoryOrPreselected(o1);
            final boolean o2selected = isMandatoryOrPreselected(o2);
            if (o1selected ^ o2selected) {
                return o1selected ? -1 : 1;
            }

            return o1.compareTo(o2);
        });
        //scroll to top
        if (fileChooser.getComponent() instanceof final JBTable table) {
            table.changeSelection(0, 0, false, false);
        }
    }

    @Override
    protected void setList(final List<ModuleDescriptor> allElements) {
        getContext().setHybrisModulesToImport(allElements);
    }

    @Override
    public void open(@Nullable final HybrisProjectSettings settings) throws ConfigurationException {
        refresh(settings);
    }

    @Override
    public void refresh(final HybrisProjectSettings settings) {
        try {
            final var filteredModuleToImport = getContext().getBestMatchingExtensionsToImport(settings);
            this.getContext().setList(filteredModuleToImport);
        } catch (ConfigurationException e) {
            // no-op already validated
        }
    }

    @Override
    protected boolean isElementEnabled(final ModuleDescriptor yModuleDescriptor) {
        if (yModuleDescriptor instanceof ConfigModuleDescriptor && yModuleDescriptor.isPreselected()) {
            return false;
        }
        if (yModuleDescriptor instanceof PlatformModuleDescriptor) {
            return false;
        }
        if (yModuleDescriptor instanceof YPlatformExtModuleDescriptor) {
            return false;
        }

        return super.isElementEnabled(yModuleDescriptor);
    }

    @Override
    @Nullable
    protected Icon getElementIcon(final ModuleDescriptor item) {
        if (this.isInConflict(item)) {
            return AllIcons.Actions.Cancel;
        }
        if (item instanceof YCustomRegularModuleDescriptor) {
            return HybrisIcons.EXTENSION_CUSTOM;
        }
        if (item instanceof ConfigModuleDescriptor) {
            return HybrisIcons.EXTENSION_CONFIG;
        }
        if (item instanceof PlatformModuleDescriptor) {
            return HybrisIcons.EXTENSION_PLATFORM;
        }
        if (item instanceof YPlatformExtModuleDescriptor) {
            return HybrisIcons.EXTENSION_EXT;
        }
//        if (item instanceof YOotbRegularModuleDescriptor) {
//            return HybrisIcons.EXTENSION_OOTB;
//        }
        if (item instanceof YWebSubModuleDescriptor) {
            return HybrisIcons.EXTENSION_WEB;
        }
        if (item instanceof YCommonWebSubModuleDescriptor) {
            return HybrisIcons.EXTENSION_COMMON_WEB;
        }
        if (item instanceof YAcceleratorAddonSubModuleDescriptor) {
            return HybrisIcons.EXTENSION_ADDON;
        }
//        if (item instanceof YBackofficeSubModuleDescriptor) {
//            return HybrisIcons.EXTENSION_BACKOFFICE;
//        }
//        if (item instanceof YBackofficeSubModuleDescriptor) {
//            return HybrisIcons.EXTENSION_BACKOFFICE;
//        }
//        if (item instanceof YHacSubModuleDescriptor) {
//            return HybrisIcons.EXTENSION_HAC;
//        }
//        if (item instanceof YHmcSubModuleDescriptor) {
//            return HybrisIcons.EXTENSION_HMC;
//        }

        return HybrisIcons.Y_LOGO_BLUE;
    }

    private boolean isMandatoryOrPreselected(final ModuleDescriptor descriptor) {
        return descriptor.getImportStatus() == MANDATORY || descriptor.isPreselected();
    }

    private boolean isPlatformExtDescriptor(final ModuleDescriptor descriptor) {
        return descriptor instanceof YPlatformExtModuleDescriptor
            || descriptor instanceof PlatformModuleDescriptor;
    }

    private boolean isCustomDescriptor(final ModuleDescriptor descriptor) {
        return descriptor instanceof YCustomRegularModuleDescriptor
            || descriptor instanceof ConfigModuleDescriptor
            || (descriptor instanceof final YSubModuleDescriptor ySubModuleDescriptor && ySubModuleDescriptor.getOwner() instanceof YCustomRegularModuleDescriptor);
    }

    private void markSubmodules(final YModuleDescriptor yModuleDescriptor, final boolean marked) {
        yModuleDescriptor.getSubModules().forEach(
            it -> fileChooser.setElementMarked(it, marked)
        );
    }
}

File: src/com/intellij/idea/plugin/hybris/project/wizard/SelectHybrisModulesToImportStep.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019-2023 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.project.wizard

import com.intellij.ide.util.ElementsChooser
import com.intellij.ide.util.projectWizard.WizardContext
import com.intellij.idea.plugin.hybris.common.utils.HybrisIcons
import com.intellij.idea.plugin.hybris.project.descriptors.ModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.ModuleDescriptorImportStatus
import com.intellij.idea.plugin.hybris.project.descriptors.YModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.YSubModuleDescriptor
import com.intellij.idea.plugin.hybris.project.descriptors.impl.*
import com.intellij.idea.plugin.hybris.settings.HybrisProjectSettings
import com.intellij.openapi.options.ConfigurationException
import com.intellij.ui.table.JBTable
import org.apache.commons.lang3.BooleanUtils

class SelectHybrisModulesToImportStep(wizard: WizardContext) : AbstractSelectModulesToImportStep(wizard), OpenSupport, RefreshSupport {

    private var selectionMode = ModuleDescriptorImportStatus.MANDATORY

    override fun init() {
        fileChooser.addElementsMarkListener(ElementsChooser.ElementsMarkListener<ModuleDescriptor> { element, isMarked ->
            if (element is YModuleDescriptor) {
                if (isMarked) {
                    element.getAllDependencies()
                        .filterNot { BooleanUtils.isNotFalse(fileChooser.elementMarkStates[it]) }
                        .forEach {
                            fileChooser.setElementMarked(it, true)
                            if (selectionMode === ModuleDescriptorImportStatus.MANDATORY) {
                                it.importStatus = ModuleDescriptorImportStatus.MANDATORY
                            }
                        }
                }

                // Re-mark sub-modules accordingly
                markSubmodules(element, isMarked)
            }

            fileChooser.repaint()
        })
    }

    override fun updateStep() {
        context.setCoreStepModuleList()

        super.updateStep()

        selectionMode = ModuleDescriptorImportStatus.MANDATORY

        for (index in 0 until fileChooser.elementCount) {
            val yModuleDescriptor = fileChooser.getElementAt(index)
            if (yModuleDescriptor.isPreselected()) {
                fileChooser.setElementMarked(yModuleDescriptor, true)
                yModuleDescriptor.importStatus = ModuleDescriptorImportStatus.MANDATORY
            }
        }

        selectionMode = ModuleDescriptorImportStatus.UNUSED

        val duplicateModules: MutableSet<String> = HashSet()
        val uniqueModules: MutableSet<String> = HashSet()

        context.list
            ?.forEach {
                if (uniqueModules.contains(it.name)) {
                    duplicateModules.add(it.name)
                } else {
                    uniqueModules.add(it.name)
                }
            }

        // TODO: improve sorting
        fileChooser.sort { o1: ModuleDescriptor, o2: ModuleDescriptor ->
            val o1dup = duplicateModules.contains(o1.name)
            val o2dup = duplicateModules.contains(o2.name)
            if (o1dup xor o2dup) {
                return@sort if (o1dup) -1 else 1
            }
            val o1custom = isCustomDescriptor(o1)
            val o2custom = isCustomDescriptor(o2)
            if (o1custom xor o2custom) {
                return@sort if (o1custom) -1 else 1
            }

            // de-boost mandatory Platform extensions
            val o1ext = isPlatformExtDescriptor(o1)
            val o2ext = isPlatformExtDescriptor(o2)
            if (o1ext xor o2ext) {
                return@sort if (o2ext) -1 else 1
            }
            val o1selected = isMandatoryOrPreselected(o1)
            val o2selected = isMandatoryOrPreselected(o2)
            if (o1selected xor o2selected) {
                return@sort if (o1selected) -1 else 1
            }
            o1.compareTo(o2)
        }
        //scroll to top
        (fileChooser.component as? JBTable)
            ?.changeSelection(0, 0, false, false)
    }

    override fun setList(allElements: MutableList<ModuleDescriptor>?) {
        context.setHybrisModulesToImport(allElements)
    }

    override fun open(settings: HybrisProjectSettings?) {
        refresh(settings)
    }

    override fun refresh(settings: HybrisProjectSettings?) {
        try {
            val filteredModuleToImport = context.getBestMatchingExtensionsToImport(settings)
            context.setList(filteredModuleToImport)
        } catch (e: ConfigurationException) {
            // no-op already validated
        }
    }

    override fun isElementEnabled(element: ModuleDescriptor?) = when {
        element is PlatformModuleDescriptor -> false
        element is YPlatformExtModuleDescriptor -> false
        element is ConfigModuleDescriptor && element.isPreselected() -> false
        else -> super.isElementEnabled(element)
    }

    override fun getElementIcon(item: ModuleDescriptor?) = when {
        item == null -> HybrisIcons.Y_LOGO_BLUE
        isInConflict(item) -> HybrisIcons.CANCEL
        item is YCustomRegularModuleDescriptor -> HybrisIcons.EXTENSION_CUSTOM
        item is ConfigModuleDescriptor -> HybrisIcons.EXTENSION_CONFIG
        item is PlatformModuleDescriptor -> HybrisIcons.EXTENSION_PLATFORM
        item is YPlatformExtModuleDescriptor -> HybrisIcons.EXTENSION_EXT
        item is YOotbRegularModuleDescriptor -> HybrisIcons.EXTENSION_OOTB
        item is YWebSubModuleDescriptor -> HybrisIcons.EXTENSION_WEB
        item is YCommonWebSubModuleDescriptor -> HybrisIcons.EXTENSION_COMMON_WEB
        item is YAcceleratorAddonSubModuleDescriptor -> HybrisIcons.EXTENSION_ADDON
        item is YBackofficeSubModuleDescriptor -> HybrisIcons.EXTENSION_BACKOFFICE
        item is YHacSubModuleDescriptor -> HybrisIcons.EXTENSION_HAC
        item is YHmcSubModuleDescriptor -> HybrisIcons.EXTENSION_HMC
        else -> HybrisIcons.Y_LOGO_BLUE
    }

    private fun isMandatoryOrPreselected(descriptor: ModuleDescriptor) = descriptor.importStatus === ModuleDescriptorImportStatus.MANDATORY
        || descriptor.isPreselected()

    private fun isPlatformExtDescriptor(descriptor: ModuleDescriptor) = descriptor is YPlatformExtModuleDescriptor
        || descriptor is PlatformModuleDescriptor

    private fun isCustomDescriptor(descriptor: ModuleDescriptor) = descriptor is YCustomRegularModuleDescriptor
        || descriptor is ConfigModuleDescriptor
        || (descriptor is YSubModuleDescriptor && descriptor.owner is YCustomRegularModuleDescriptor)

    private fun markSubmodules(yModuleDescriptor: YModuleDescriptor, marked: Boolean) {
        yModuleDescriptor.getSubModules()
            .forEach { fileChooser.setElementMarked(it, marked) }
    }
}